File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;
class Solution {
    public int[] solution(int[] numbers) {
        Stack<Pair> st = new Stack<>();
        int[] answer = new int[numbers.length];
        for(int i=0;i<numbers.length;i++){
            while(!st.isEmpty() && st.peek().first<numbers[i]){
                answer[st.peek().second] = numbers[i];
                st.pop();
            }
            st.push(new Pair(numbers[i],i));
        }
        while(!st.isEmpty()){
            answer[st.peek().second] = -1;
            st.pop();
        }
        return answer;
    }
    
    class Pair{
        int first;
        int second;
        public Pair(int first, int second){
            this.first = first;
            this.second = second;
        }
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;
stack<pair<int,int>>st;
vector<int> solution(vector<int> numbers) {
    vector<int> answer;
    answer.resize(numbers.size());
    for(int i=0;i<numbers.size();i++){
        while(!st.empty()&&st.top().first<numbers[i]){
            answer[st.top().second] = numbers[i];
            st.pop();
        }
        st.push({numbers[i],i});
    }
    while(!st.empty()){
        answer[st.top().second] = -1;
        st.pop();
    }
    return answer;
}